### ISSUE TITLE
Screen magnifier / zoom tool?

### ISSUE THREAD

**Issue by spl237** [labwc/labwc#1703]
One thing which we have provided on RPi Desktop for several years now is a screen magnifier to enable users with vision difficulties to enlarge parts of the screen.

Under X, this was a separate magnifier application which provided a magnified area around the mouse pointer - https://github.com/raspberrypi-ui/magnifier

This obviously could not be used on Wayland, so under wayfire we used a slightly modified version of the standard zoom plugin which simply magnified the entire screen centred on the mouse pointer - https://github.com/WayfireWM/wayfire/blob/master/plugins/single_plugins/zoom.cpp

Is there any way to do something similar in labwc? I don't know what proportion of our user base ever made use of either feature - not many, I suspect - but it would be nice to be able to retain similar functionality for those who do, if at all possible. I'm not sure of the best way to go about implementing it, though.

(At a quick glance, the Wayfire zoom tool looks to be doing things in a not-very-wlroots fashion, so I suspect would not be appropriate for labwc.)

**Comment by redtide** [comment]
This reminds me [hyprpicker](https://github.com/hyprwm/hyprpicker)

**Comment by Consolatis** [comment]
There is the approach taken by tools like https://github.com/SIMULATAN/hyprmag (never tested it so I can't say anything about its quality). Those appear to take a screenshot, showing that one in fullscreen and then magnifying things around the cursor. I am not sure if those allow clicking on items through the lens though, it \*might\* be possible to archive that when manipulating the input region but I never tested that approach. With the an empty input region it could theoretically also be possible to just use a transparent layershell overlay and only do screenshots of the area around the cursor and then magnify those but I am not sure if the layershell client would even get the mouse events with an empty input region.

I am not quire sure how we could do something on the compositor side when using scene graph.

**Comment by spl237** [comment]
That looks promising, certainly - I'll give it a try. Ideally we do need people to be able to interact with the magnified area, at least at the level of mouse clicks being handled. But anything is better than nothing!

**Comment by droc12345** [comment]
Wayfire was the first wayland compositor that I used. Even today it's my backup.

The zoom was nice, you could zoom on some part of the wallpaper or specific windows. 
I set the zoom to the scroll wheel, nice to be able to scroll up to what size is needed to make it readable.

**Comment by droc12345** [comment]
I know with wayfire, effectively the whole view is zoomed, I used it to play pysol on my 4k monitor,
since it used a set size font that was way too small normally. 

Not sure how hyprmag does it, as I haven't messed with hyprland in months.

**Comment by Consolatis** [comment]
A compositor can do all kind of things if it renders the screen content itself, however since labwc 0.6.0 we don't do that anymore and delegate the actual rendering and damage detection to the wlroots scene graph API. And that one doesn't expose anything that we could use here AFAIR.

**Comment by spl237** [comment]
I think that while the circular magnifying lens in the likes of hypermag is a nice visual effect, the wayfire approach of zooming the entire view and then showing a window into it the size of the monitor is actually more useful for a visually-impaired user - that would be my preferred approach, and I suspect it would be easier to do.

**Comment by Consolatis** [comment]
I don't think labwc can implement something like this properly with the current state of the scene graph API.
The external application idea could work to an extent (only operate on a stale screenshot created at the start of the operation, no ability to pass through clicks to the underlying surface).

**Comment by Consolatis** [comment]
> * Create a no-input, transparent surface that covers the entire output so we can get the global cursor position

That is the issue, at least on my tests I didn't get cursor movements with an empty input region.

**Comment by Consolatis** [comment]
Had some fun with a labwc internal proof-of-concept: https://github.com/Consolatis/labwc/tree/feature/magnifier

![20240411_16h20m22s_grim](https://github.com/labwc/labwc/assets/35009135/edd5446e-a4e9-4c4b-bb58-4ea2a1168603)

It only works properly when using `WLR_SCENE_DEBUG_DAMAGE=rerender WLR_NO_HARDWARE_CURSORS=1 WLR_RENDERER=pixman labwc` and I only tested it nested.

We might overcome the damage issue and potentially the non-hardware cursor issue but I have absolutely no idea how we could overcome the pixman requirement. It might be possible by using our own render_pass after the wlr_scene one but I didn't look into that so far.

The quality is also not great (as expected when just scaling lower resolutions up).

**Comment by spl237** [comment]
That looks really promising. I think we (RPi) might be find with using pixman, and I think we already need to disable h/w cursors in some circumstances, so I'll have a play when I get some free time. Excellent work - many thanks!

**Comment by spl237** [comment]
One other idea - I'm not sure exactly how the likes of wayvnc get access to what is on the screen, but might that be an alternative approach?

**Comment by Consolatis** [comment]
> One other idea - I'm not sure exactly how the likes of wayvnc get access to what is on the screen, but might that be an alternative approach?

The wlr-screencopy protocol. It is what the other external tools are using, with the issues as described above.

**Comment by spl237** [comment]
@Consolatis, I've finally had some time to play with this, and it's great! Does exactly what I need in terms of a magnifier. I'm currently tweaking hotkeys and mouse bindings so that it is controlled in the same way as our existing magnifiers, but once that is done, it'll be perfect for our needs.

Is this the sort of thing that might get merged into the core labwc code? If not, we (Raspberry Pi) could maintain it as a local patch.

**Comment by Consolatis** [comment]
> Is this the sort of thing that might get merged into the core labwc code? If not, we (Raspberry Pi) could maintain it as a local patch.

Personally I'd be fine with that if its by default disabled. It is also pretty self contained and thus the impact on the rest of the codebase should be minimal. Providing accessibility options is important IMHO.

**Comment by spl237** [comment]
OK, in that case I will work it up into a PR, if that is ok with you?

**Comment by Consolatis** [comment]
> OK, in that case I will work it up into a PR, if that is ok with you?

Sure, feel free to squash and modify the commits as you see fit.
